Database Migration (#275)
reflectoring/reflectoring.github.io: _posts/2019-12-16-database-migration-in-self-contained-systems.md

---

title: Database Migration in Self Contained Systems
categories: [craft]
modified: 2019-12-16
author: artur
excerpt: "A tutorial on how to migrate databases in SCS"
image:
  auto: 0066-database-migration
---

> When you develop a application nowadays, you as programmer just want to get the database and 

Reviewer: Try to use "we" instead of "you" throughout the text. That makes for a more conversational tone.

Author: done

to add new scripts without thinking about database maintenance. This problem is already solved
by frameworks like [flyway](https://flywaydb.org/) or [liquibase](https://www.liquibase.org/). 

> The database is provided to you in this case or you have to take care about this.

Reviewer: I don't understand the value of this sentence. Please either elaborate or delete it.

Author: rewritten

> If you develop a self contained system (SCS), it could have a database within. In this 

Reviewer: Please replace "self contained" with "self-contained" throughout the text.

Author: done

> If you develop a self contained system (SCS), it could have a database within. In this 

Reviewer: ```suggestion
If you develop a self contained system (SCS), it probably has it's own database. In this 
```

Author: replaced by other text

> article we are going to have a look how to deliver a database with the SCS and how to have the migration

Reviewer: ```suggestion
article we are going to have a look at how to deploy a database together with the SCS and how to control database migration.
```

Author: replaced by other text

> control over the database and not over the business data only.          

Reviewer: ```suggestion      
```

Author: rewritten

{% include github-project.html url="https://github.com/arkuksin/scs-db-versioning" %}

## General Approach

> First we create the database and set the admin user by creating the database. Then we create database

Author: done

configuration for the application, that we working on. It means creation a scheme, a user, privileges and so on.
After that we start the application and connect the database with created configuration. Let's look how can we automate 
this. I created a example project to show this in action.

## Project Structure

> This example is based on three technologies:

Author: done

* Docker as container engine

> * Kubernetes as container orchestration system

Author: described in introduction

* Spring Boot as our application  

> The project is a SCS with a [Postgresql](https://www.postgresql.org/) Database and a very simple

Author: PostgreSQL is correct.

Spring Boot Application called `User Service`. According to these components the project consists of two parts:
* `k8s` folder with Kubernetes manifest files
* `src` with source code for `User Service`

A build pipeline is supposed to exist and it should 
* build the application as a Docker image,
* publish the image to a Private Docker Registry,
* apply Kubernetes manifest files,

but we let the pipeline out of scope in this article.    
 
## Kubernetes Objects
 

> Let's look at Kubernetes Configuration. 

Author: done

 
### Base Configuration
 

>  In the folder ``k8s/base`` we create a `ConfigMap` and `Secret` with Connection Properties for the `User Service`.

Author: done

>  It looks like 

Author: deleted

>  ConfigMap

Author: I linked and described the Kubernetes Object

 ````yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-configmap
  namespace: migration
data:
  spring.datasource.username: user_service
  spring.datasource.driver-class-name: org.postgresql.Driver
  spring.datasource.url: jdbc:postgresql://postgres:5432/user_service
````
Secret
````yaml
apiVersion: v1
kind: Secret
metadata:
  name: user-secret
  namespace: migration
type: Opaque
data:
  spring.datasource.password: bXlfc2VydmljZV9wYXNzd29yZA== #my_service_password
````     

These data are used in two cases:
   * creation of database user for the application,
   * connection database from the application.
   

> The point is that we have the single source of data in both cases.

Author: removed

### Database

> With the scripts from ``k8s/postgres`` we create a Postgresql Database as `StatefulSet` with an admin user. The password for

Author: I described the approach in the introduction

the user is read from another secret:

`````yaml
apiVersion: v1
kind: Secret
metadata:
  name: postgres
  namespace: migration
type: Opaque
data:
  password: bXlzZWNyZXQ= #mysecret
````` 

This password is needed later to access the database for setting our configuration
 for ``User Service``.

### Migration

> Now let's look at the most interesting part of kubernetes configuration.

Author: done

> Our goal is to automate the creation of schema,

Author: done

users, privileges and so on. We can create all of this with SQL scripts.

> Fortunately we can use [flyway](https://flywaydb.org/) to migrate SQL scripts 

Author: Flyway is correct

as we know it by migration from an application!

> First we create a ``ConfigMap`` with the SQL scripts

Author: I linked and described it

````yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: postgres-configmap
  namespace: migration
data:
  V1_1__create_user.sql: |
    CREATE USER ${username} WITH
      LOGIN
      NOSUPERUSER
      NOCREATEDB
      NOCREATEROLE
      INHERIT
      NOREPLICATION
      CONNECTION LIMIT -1
      ENCRYPTED PASSWORD '${password}';
  V1_2__create_db.sql: |
    CREATE DATABASE ${username} WITH
      OWNER = ${username}
      ENCODING  = 'UTF8'
      CONNECTION LIMIT = -1;
  V1_3__grant_privileges.sql: |
    GRANT ALL ON DATABASE ${username} TO ${username}

> ````     

Author: explained

Next we create a ``Job``, that migrates these scripts into the database. Since we use Kubernetes,
we can use [flyway docker container](https://hub.docker.com/r/boxfuse/flyway/)

> to start the migration. The `Job` mounts the `ConfigMap`

Author: explained

              value: jdbc:postgresql://postgres:5432/postgres
            - name: FLYWAY_USER
              value: admin
            - name: FLYWAY_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: postgres
                  key: password
            - name: FLYWAY_PLACEHOLDER_REPLACEMENT
              value: "true"
            - name: FLYWAY_PLACEHOLDERS_USERNAME
              valueFrom:
                configMapKeyRef:
                  name: user-configmap
                  key: spring.datasource.username
            - name: FLYWAY_PLACEHOLDERS_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: user-secret
                  key: spring.datasource.password
          volumeMounts:
            - mountPath: /flyway/sql
              name: sql
      volumes:
        - name: sql
          configMap:
            name: postgres-configmap
      restartPolicy: Never
````

So the ``ConfigMap`` is mounted to the path `/flyway/sql`, which is the default path for SQL migration
with flyway. Environment variables ``FLYWAY_USER`` and `FLYWAY_PASSWORD` define the user for connecting the database,

> who is the admin user. The variable ``FLYWAY_PLACEHOLDER_REPLACEMENT`` enables the placeholder replacement.

Author: done

The variable ``FLYWAY_PLACEHOLDERS_USERNAME`` is set to the username of `User Service` und the variable
``FLYWAY_PLACEHOLDERS_PASSWORD`` to its password.

After these scripts have been running, the database for the user, the user und the privileges are created and are under
flyway control. So if you want to change you database configuration, you have to add a new SQL script to the 

> ``ConfigMap`` and start the `Job` again. Ideally the pipeline does it. You just have to push the changes of 

Author: I tried to explain it

`ConfigMap` to your repository.

That always works, when you want to set up new environments with your SMSs. The SCM will create
the database and configure it for the application automatically.

If you have many environments in your delivery pipeline you have configure it once and it works on every
environment.

### Deployment of Application
To access the database from the application, it has to read the database configuration from ``ConfigMap``
and ``Secret``. Of course there is a `spring-boot-starter` for this goal.

These dependencies should be added to your ``build.gradle``

````groovy
    implementation 'org.springframework.cloud:spring-cloud-starter-kubernetes-config:1.0.4.RELEASE'
    implementation 'org.springframework.cloud:spring-cloud-starter-kubernetes:1.0.4.RELEASE'
```` 

> Then we have to reference the names of the ``ConfigMap`` and `Secret` in `bootstrap.yml`. Not `application.yml`!

Author: done

````yaml
spring:
  application:
    name: user-service
  cloud:
    kubernetes:
      config:
        name: user-configmap
      secrets:
        name: user-secret
````

That's it. The values

 * `` spring.datasource.username``
 * `` spring.datasource.driver-class-name ``
 * ``spring.datasource.url``
 * `` spring.datasource.password``
 
 are read from Kubernetes Object and Spring Boot can use them. The same values were used by the ``Job``
 to create a user in the database. This application starts its database migration too and uses 
 the flyway tool for that again, but this time for migrating business data. It is easily made by adding a dependency
 to ``build.gradle``
 
 ````groovy
implementation 'org.flywaydb:flyway-core'
```` 

## Notes
* I used [minikube](https://github.com/kubernetes/minikube) to run the SCS locally.
* For more security don't put secrets in plain to the files in repository.

## Conclusion

> There is a way to deliver the database with self contained system and have version control over

Author: introduction and conclusion are rewritten

to set up an empty database for step 1. Also, we can easily deploy an application as a docker container for step 3. But how can we automatically
configure the database and how can we control different versions for this configuration?

Using the declarative approach of Kubernetes as a container orchestration system, we can easily build up the automated process to deliver the SCS.
We can just declare the desired state of components like a database or application and Kubernetes takes care of the rest.
But it's not possible to declare the desired database configuration with its schemas, users, privileges and so on, because it is the *internal* configuration
of the database.

Imagine we have a pipeline to deliver our SCS. This pipeline does the three steps from the list above.
When we start the pipeline for the first time, the database will be created and configured and the application will be deployed.
If we later make a change in the business logic, the pipeline will run these steps and Kubernetes will not detect any changes on the database, but on the business logic application.
In this case, Kubernetes is responsible for updating our application. It works fine with Kubernetes. 

> **But we also want to be able to make changes to the database configuration (e.g. to change some permissions) and have that change deploy automatically**.

Author: > (e.g. to change some permissions)

this is fine, but I just want to show in this comment another example from my experience. I added a quartz framework to my application and I wanted to add a new scheme for that to split the business data and quartz data. This was a good use case to apply the approach from this article.

  spring.datasource.driver-class-name: org.postgresql.Driver
  spring.datasource.url: jdbc:postgresql://postgres:5432/post_service
```

A `Secret` is a Kubernetes Object for storing sensitive data. Similar to a `ConfigMap`, a `Secret` can also be read at runtime: 

```yaml
# base/secret.yml
apiVersion: v1
kind: Secret
metadata:
  name: post-secret
  namespace: migration
type: Opaque
data:
  spring.datasource.password: bXlfc2VydmljZV9wYXNzd29yZA==
```    

As we can see, the data from `post-configmap` and the password from `post-secret` are the data for connecting to the database
with the user `post_service`. After deploying the Spring Boot application, our application and database should be connected and ready to use.

### Database

With the scripts from `k8s/postgres` we create a PostgreSQL Database as a
[StatefulSet](https://cloud.google.com/kubernetes-engine/docs/concepts/statefulset) with an admin user.
A `StatefulSet` is a Kubernetes Object that defines components with unique, persistent identities 

> and with persistent disk storage. A database fits well into the use case of a `StatefulSet`. 

Author: Does it make sense to add the `yaml` code for the `StatefulSet` to complete this example? I didn't add it in my version, because it is much code and it is not more than just the creation of a PostgreSQL database.
